beacon-chain/forkchoice/protoarray/optimistic_sync.go

```diff
@@ -88,3 +88,204 @@ func (f *ForkChoice) Optimistic(ctx context.Context, root [32]byte, slot types.S
+	f.syncedTips.Lock()
+	defer f.syncedTips.Unlock()
+	_, ok = f.syncedTips.validatedTips[root]
+	if ok {
+		return nil
+	}
+
+	// Cache root and slot to validated tips
+	f.syncedTips.validatedTips[root] = node.slot
+
+	// Compute the full valid path from the given node to its previous synced tip
+	// This path will now consist of fully validated blocks. Notice that
+	// the previous tip may have been outside the Fork Choice store.
+	// In this case, only one block can be in syncedTips as the whole
+	// Fork Choice would be a descendant of this block.
+	validPath := make(map[uint64]bool)
+	for {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		parentIndex := node.parent
+		if parentIndex == NonExistentNode {
+			break
+		}
+		if parentIndex >= uint64(len(f.store.nodes)) {
+			return errInvalidNodeIndex
+		}
+		node = f.store.nodes[parentIndex]
+		_, ok = f.syncedTips.validatedTips[node.root]
+		if ok {
+			break
+		}
+		validPath[parentIndex] = true
+	}
+
+	// Retrieve the list of leaves in the Fork Choice
+	// These are all the nodes that have NonExistentNode as best child.
+	leaves, err := f.store.leaves(ctx)
+	if err != nil {
+		return err
+	}
+
+	// For each leaf, recompute the new tip.
+	newTips := make(map[[32]byte]types.Slot)
+	for _, i := range leaves {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		node = f.store.nodes[i]
+		j := i
+		for {
+			// Stop if we reached the previous tip
+			_, ok = f.syncedTips.validatedTips[node.root]
+			if ok {
+				newTips[node.root] = node.slot
+				break
+			}
+
+			// Stop if we reach valid path
+			_, ok = validPath[j]
+			if ok {
+				newTips[node.root] = node.slot
+				break
+			}
+
+			j = node.parent
+			if j == NonExistentNode {
+				break
+			}
+			if j >= uint64(len(f.store.nodes)) {
+				return errInvalidNodeIndex
+			}
+			node = f.store.nodes[j]
+		}
+	}
+
+	f.syncedTips.validatedTips = newTips
+	return nil
+}
+
+// This function removes a node that has become INVALID according to the
+// EL. It updates the synced_tips to reflect this change.
+func (f *ForkChoice) Invalid(ctx context.Context, root [32]byte) error {
+	f.store.nodesLock.Lock()
+	defer f.store.nodesLock.Unlock()
+	idx, ok := f.store.nodesIndices[root]
+	if !ok {
+		return errInvalidNodeIndex
+	}
+	node := f.store.nodes[idx]
+	parentIndex := node.parent
+
+	f.syncedTips.Lock()
+	defer f.syncedTips.Unlock()
+
+	// This should not happen
+	if parentIndex == NonExistentNode {
+		return nil
+	}
+	parent := f.store.nodes[parentIndex]
+	parentRoot := parent.root
+
+	// Check if after removing the child node, we need to update the
+	// parent's best child and descendant.
+	needsUpdateParentsChildren := false
+	if parent.bestChild == idx || parent.bestDescendant == idx {
+		needsUpdateParentsChildren = true
+	}
+
+	// delete the invalid node, order is important
+	f.store.nodes = append(f.store.nodes[:idx], f.store.nodes[idx+1:]...)
+	delete(f.store.nodesIndices, root)
+	// Fix best parent and best child for each node
```

Comment:
```suggestion
	// Fix parent and best child for each node
```

there's just one parent per node